### ISSUE TITLE
Version 2.8.1 with .NET 4.7.1 creates dozens of unneeded library references

### ISSUE THREAD

**Issue by jmhongithub** [OfficeDev/Open-XML-SDK#401]
**Description**

If you add version 2.8.1 (via NuGet) to a project that targets .NET 4.7.1, then it will pull down dozens of System.X libraries, some of which will causes reference issues with larger projects. 

**Information**

- .NET Target: 4.7.1
- DocumentFormat.OpenXml Version: 2.8.1

**Repro**
Create a new solution with a console app project targeting 4.7.1. Add version 2.8.1 via NuGet. 

**Observed**

Dozens of extra unneeded libraries are added. 

**Expected**

It should only add 3 additional libraries: 
* Microsoft.NETCore.Platforms
* NETStandard.Library
* System.IO.Packaging

**Workaround**
If you install NETStandard.Library before DocumentFormat.OpenXml, and upgrade Microsoft.NETCore.Platforms, then only additional library installed is System.IO.Packaging.

**Comment by huysentruitw** [comment]
Everything pre netstandard should just reference WindowsBase IMO.

**Comment by twsouthwick** [comment]
@jmhongithub What kind of project are you making? Is the net effect of your workaround that you have less assemblies in the output, or is it just that adding DocumentFormat.OpenXml now doesn't say it's adding libraries?

Most of these extra libraries are actually thin wrappers that are facades to other libraries. Unfortunately, we're stuck with them until a good linker can be produced that will strip those away.


As an example, consider the following scenario:

1. Create a .NET Standard 1.3 library with no other platform support. Let's call this `StandardLib`
2. Add a reference to `DocumentFormat.OpenXml` to `ClassLib1`
4. Add the following method:
```csharp
using DocumentFormat.OpenXml.Packaging;
using System.IO.Packaging;

namespace ClassLibrary1
{
    public class Class1
    {
        public static void Run(Package package)
        {
            WordprocessingDocument.Create(package, DocumentFormat.OpenXml.WordprocessingDocumentType.Document);
        }
    }
}
```
3. Create a .NET Framework 4.6 application. Let's call this `FrameworkApp`
4. Reference `StandardLib` in `FrameworkApp`
5. Add the following to the `FrameworkApp`:
```csharp
using System.IO;
using System.IO.Packaging;

namespace ConsoleApp4
{
    class Program
    {
        static void Main(string[] args)
        {
            var path = Path.GetTempFileName();
            using (var package = Package.Open(path, FileMode.CreateNew, FileAccess.ReadWrite))
            {
                ClassLibrary1.Class1.Run(package);
            }
        }
    }
}
```

If we were to reference `WindowsBase` in the .NET 4.6 build of DocumentFormat.OpenXml, then you'll end up with the following error:

```
The type 'Package' is defined in an assembly that is not referenced. You must add a reference to assembly 'System.IO.Packaging, Version=4.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

There's not much we can do that I know of in the SDK to minimize this without introducing more issues than we fix. A possible solution is to remove all the public surface from the .NET Standard build that references `System.IO.Packaging`, but I'm not sure how feasible that is.

**Comment by twsouthwick** [comment]
@huysentruitw The issue you're describing is likely something different as it's issues with runtime resolution. It's potentially the same root cause, though, but can you start another issue to discuss that, and we can merge it if it is the same?

**Comment by twsouthwick** [comment]
@jmhongithub Can you respond to the questions? Unless I hear otherwise, this will be closed as "by-design" since I believe the issue you are describing is that when targeting .NET 4.6 a number of assembly shims are brought down. This is nothing to do with the SDK, but the way that .NET currently works.

**Comment by jmhongithub** [comment]
@twsouthwick,  I'm using it in stand alone project that is referenced by an ASP.NET MVC application. The net effect is adding 40+ unneeded libraries to the project. One of those projects happen to cause a conflict with another another I was already using. The other impact is that I'm in environment where the build server isn't allowed to have a direct connection to the Internet. This means uploading all those packages to local NuGet repo.

Again I documented a workaround, but the default behavior of referencing the last version of OpenXML is not ideal.

**Comment by twsouthwick** [comment]
So I'm identifying two issues you're seeing:

1. You are having a conflict with another library when trying to upgrade to 2.8.1 What library is that? What errors are you getting?
2. As for uploading the packages to a local NuGet repo, are you using a hosted NuGet server? MyGet and others allow for the ability to "mirror" packages and all their dependencies that should get around that. This doesn't help, though, if you are using a file share.

Most likely the issue 1 is caused by not correctly including transitive dependencies since you're probably using packages.config with ASP.NET and it's easy to get out of sync (both in packages.config and the web.config for binding redirects). Unfortunately, if you are using EF or bringing in any JS libraries via nuget, you can't switch to PackageReference which will automatically sort that out for you, and you'll just include the library you care about with worrying about the libraries it may bring in.

**Comment by twsouthwick** [comment]
Since this is how the ecosystem works now with .NET Standard, there's nothing we can do while still maintaining support for all platforms. I'm going to close this issue. If you feel that it wasn't addressed, please reopen and provide more details.

**Comment by yurriiy** [comment]
When i add 2.8.1 (via NuGet) to a project that targets .NET 4.6.2, i see references:
> Microsoft.Win32.Primitives.dll
> System.AppContext.dll
> System.Console.dll
> System.Diagnostics.DiagnosticSource.dll
> System.Diagnostics.DiagnosticSource.xml
> System.Diagnostics.Tracing.dll
> System.Globalization.Calendars.dll
> System.IO.dll
> System.IO.Compression.dll
> System.IO.Compression.ZipFile.dll
> System.IO.FileSystem.dll
> System.IO.FileSystem.Primitives.dll
> System.IO.Packaging.dll
> System.Net.Http.dll
> System.Net.Sockets.dll
> System.Reflection.dll
> System.Runtime.dll
> System.Runtime.Extensions.dll
> System.Runtime.InteropServices.dll
> System.Runtime.InteropServices.RuntimeInformation.dll
> System.Security.Cryptography.Algorithms.dll
> System.Security.Cryptography.Encoding.dll
> System.Security.Cryptography.Primitives.dll
> System.Security.Cryptography.X509Certificates.dll
> System.Xml.ReaderWriter.dll

Can i simply add reference to DocumentFormat.OpenXml.dll for correct work?

**Comment by jmhongithub** [comment]
@yurriiy , with 4.7.1, I found only three were needed:

- Microsoft.NETCore.Platforms
- NETStandard.Library
- System.IO.Packaging

I removed the other libraries via NuGet.

**Comment by yurriiy** [comment]
I am install, package without dependencies:
> Install-Package DocumentFormat.OpenXml -IgnoreDependencies
> Install-Package System.IO.Packaging -IgnoreDependencies
It seems that everything works, thx.

**Comment by twsouthwick** [comment]
I've submitted a PR to update System.IO.Packaging to 4.5.0 which fixes this issue and only includes the dependencies that are actually needed. Namely, this is: System.IO.Packaging and System.IO.FileSystem.Primitives.

**Comment by ackginger** [comment]
I see it's been merged, but any idea what the timeline is for their next release?

**Comment by twsouthwick** [comment]
Not sure what you mean by *their* release, but:

- `System.IO.Packaging` is released as a stable 4.5.0. You can manually update to that and get the same benefits (at least with the non-packages.config management system)
- `DocumentFormat.OpenXml` is currently in preview for 2.9.0. We hope to release a stable version of it soon

**Comment by MPasadu** [comment]
Still problems. Tried to install openxml 2.8.1 on a .net 4.7.1 project. 

- When using the standard option on dependency behavior, it will try to install 20+ dependencies. 

- When using the "highest" option on dependency behavior, it will only install System.IO.Packaing, DocumentFormat.OpenXML, and System.IO.FileSystem.Primitives (This one isn't needed). Also System.IO.FileSystem.Primitives will show the yellow warning triangle. 

The only thing that worked was @yurriiy solution:

> Install-Package DocumentFormat.OpenXml -IgnoreDependencies
> Install-Package System.IO.Packaging -IgnoreDependencies

**Comment by twsouthwick** [comment]
As my comment above says, it is fixed in 2.9.0 so installing 2.8.1 will see the same issue. We are having an issue uploading it to NuGet, but it is available via the CI feed.

System.IO.FileSystem.Primitives is necessary transitively by System.IO.Packaging.

Also, are you using `PackageReferences` (the new NuGet style) or `packages.config` (the old NuGet style)? `PackageReferences` is required to get the benefits of manually update the packages independent of the SDK.
